presto-main/src/main/java/io/prestosql/operator/unnest/BlockProducerFromSource.java

```diff
@@ -0,0 +1,337 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.unnest;
+
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.block.DictionaryBlock;
+import io.prestosql.spi.block.PageBuilderStatus;
+import io.prestosql.spi.type.Type;
+
+import java.util.Arrays;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This class is used to create an output block from a source input block.
+ * Invoking any other public method makes sense only after resetInputBlock
+ * has been invoked with a source Block.
+ *
+ * The methods append* are provided for adding elements (possibly null)
+ * to current output block.
+ *
+ * The output block can be either
+ * (1) a dictionary block created using source as the dictionary OR
+ * (2) a block created using preferred blockbuilder for type, with elements
+ *     copied over from source
+ *
+ * Dictionary output block are always preferred over copied block.
+ * A copy is performed only when:
+ * - appendNull has been invoked at-least once for current output block AND
+ * - there is not a single NULL element in the source
+ */
+public class BlockProducerFromSource
```

Comment:
Let's work on a better name. Technically, most operators produces a block from a source ;P.

I'd call out that you're essentially handling two dimensions/types of polymorphism with unnest: the type of the column being unnested and the encoding of the columns. Perhaps that can lead to a better name.

"EncodingAwareReplicator"? "DictionaryAware..."?